File: ios/LessonsPageControl.m
// Copyright 2018 David Sansome
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#import "LessonsPageControl.h"

#import "TKMSubjectChip.h"
#import "Tsurukame-Swift.h"
#import "proto/Wanikani+Convenience.h"

@interface LessonsPageControl () <TKMSubjectChipDelegate>
@end

@implementation LessonsPageControl {
  NSMutableArray<TKMSubjectChip *> *_chips;
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
  self = [super initWithCoder:aDecoder];
  if (self) {
    _chips = [NSMutableArray array];
  }
  return self;
}

#pragma mark - Layout

- (void)setSubjects:(NSArray<TKMSubject *> *)subjects {
  // Remove all existing chips.
  for (TKMSubjectChip *chip in _chips) {
    [chip removeFromSuperview];
  }
  [_chips removeAllObjects];

  // Create a chip for each subject.
  for (TKMSubject *subject in subjects) {
    TKMSubjectChip *chip = [[TKMSubjectChip alloc] initWithSubject:subject
                                                       showMeaning:false
                                                          delegate:self];
    [self addSubview:chip];
    [_chips addObject:chip];
  }

  // Create the quiz chip.
  NSAttributedString *quizText = [[NSAttributedString alloc] initWithString:@"Quiz"];
  NSArray *chipGradient = @[ (id)TKMStyle.greyColor.CGColor, (id)TKMStyle.greyColor.CGColor ];
  TKMSubjectChip *quizChip = [[TKMSubjectChip alloc] initWithSubject:nil
                                                            chipText:quizText
                                                            sideText:nil
                                                       chipTextColor:[UIColor whiteColor]
                                                        chipGradient:chipGradient
                                                            delegate:self];
  [self addSubview:quizChip];
  [_chips addObject:quizChip];

  self.currentPageIndex = _currentPageIndex;

  [self setNeedsLayout];
}

- (void)setCurrentPageIndex:(NSInteger)index {
  _currentPageIndex = index;
  for (int i = 0; i < _chips.count; ++i) {
    _chips[i].dimmed = i != index;
  }
}

- (void)layoutSubviews {
  NSArray<NSValue *> *chipFrames =
      TKMCalculateSubjectChipFrames(_chips, self.frame.size.width, NSTextAlignmentCenter);
  for (int i = 0; i < _chips.count; ++i) {
    _chips[i].frame = [chipFrames[i] CGRectValue];
  }
}

- (CGSize)sizeThatFits:(CGSize)size {
  if (!_chips.count) {
    return size;
  }
  NSArray<NSValue *> *chipFrames =
      TKMCalculateSubjectChipFrames(_chips, size.width, NSTextAlignmentCenter);
  return CGSizeMake(size.width,
                    CGRectGetMaxY([chipFrames.lastObject CGRectValue]) +
                        kTKMSubjectChipCollectionEdgeInsets.bottom);
}

#pragma mark - TKMSubjectChipDelegate

- (void)didTapSubjectChip:(TKMSubjectChip *)chip {
  for (int i = 0; i < _chips.count; ++i) {
    if (_chips[i] == chip) {
      self.currentPageIndex = i;
      [self sendActionsForControlEvents:UIControlEventValueChanged];
      break;
    }
  }
}

@end

File: ios/LessonsPageControl.swift
// Copyright 2021 David Sansome
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import Foundation

@objc
class LessonsPageControl: UIControl, SubjectChipDelegate {
  private var chips = [SubjectChip]()

  func setSubjects(_ subjects: [TKMSubject]) {
    // Remove all existing chips.
    for chip in chips {
      chip.removeFromSuperview()
    }
    chips.removeAll()

    // Create a chip for each subject.
    for subject in subjects {
      let chip = SubjectChip(subject: subject, showMeaning: false, delegate: self)
      addSubview(chip)
      chips.append(chip)
    }

    // Create the quiz chip.
    let quizText = NSAttributedString(string: "Quiz")
    let gradient: [Any] = [TKMStyle.Color.grey80.cgColor, TKMStyle.Color.grey80.cgColor]
    let quizChip = SubjectChip(subject: nil, chipText: quizText, sideText: nil,
                               chipTextColor: .white, chipGradient: gradient, delegate: self)
    addSubview(quizChip)
    chips.append(quizChip)

    currentPageIndexChanged()
    setNeedsLayout()
  }

  @objc
  var currentPageIndex = 0 {
    didSet { currentPageIndexChanged() }
  }

  private func currentPageIndexChanged() {
    for (idx, chip) in chips.enumerated() {
      chip.isDimmed = idx != currentPageIndex
    }
  }

  // MARK: - Layout

  override func layoutSubviews() {
    let frames = calculateSubjectChipFrames(chips: chips, width: frame.size.width,
                                            alignment: .center)
    for (idx, chip) in chips.enumerated() {
      chip.frame = frames[idx]
    }
  }

  override func sizeThatFits(_ size: CGSize) -> CGSize {
    if chips.isEmpty {
      return size
    }

    let frames = calculateSubjectChipFrames(chips: chips, width: frame.size.width,
                                            alignment: .center)
    return CGSize(width: size.width,
                  height: frames.last!.maxY + kSubjectChipCollectionEdgeInsets
                    .bottom)
  }

  // MARK: - SubjectChipDelegate

  func didTapSubjectChip(_ subjectChip: SubjectChip) {
    for (idx, chip) in chips.enumerated() {
      if chip == subjectChip {
        currentPageIndex = idx
        sendActions(for: .valueChanged)
        break
      }
    }
  }
}
